File: frontend/src/types/plannedCost.ts
export type PlannedCost = {
    id: string;
    tripId: string;
    name: string;
    category: PlannedCostCategory;
    amount: number;
    createdAtUtc: string;
    expenseAdded?: boolean;
    expenseId?: string | null;
};

export type PlannedCostCategory =
    | "TravelToFrom"
    | "Accommodation"
    | "LocalTransport"
    | "Food"
    | "ActivitiesAndMuseums"
    | "BarsAndNightlife"
    | "Shopping"
    | "EmergencyBuffer"
    | "Other";

export type PlannedCostForm = {
    name: string;
    category: "" | PlannedCostCategory;
    amount: string;
};

export const plannedCostCategories: Array<{
    value: PlannedCostCategory;
    label: string;
}> = [
    { value: "TravelToFrom", label: "Travel to/from" },
    { value: "Accommodation", label: "Accommodation" },
    { value: "LocalTransport", label: "Local transport" },
    { value: "Food", label: "Food" },
    { value: "ActivitiesAndMuseums", label: "Activities & museums" },
    { value: "BarsAndNightlife", label: "Bars & nightlife" },
    { value: "Shopping", label: "Shopping" },
    { value: "EmergencyBuffer", label: "Emergency buffer" },
    { value: "Other", label: "Other" },
];

/** Creates a new cost form, optionally preserving a category selected from a quick-add action. */
export const createEmptyPlannedCostForm = (
    category: PlannedCostForm["category"] = "",
): PlannedCostForm => ({
    name: "",
    category,
    amount: "",
});

File: backend/Models/PlannedCost.cs
namespace TravelAssistant.Models;

/// <summary>
/// Represents a budgeted cost before it has been spent; it may later link to one actual expense.
/// </summary>
public class PlannedCost
{
    public Guid Id { get; set; } = Guid.NewGuid();

    public Guid TripId { get; set; }

    public Trip Trip { get; set; } = null!;

    public string Name { get; set; } = "Cost item";

    public PlannedCostCategory Category { get; set; }

    public decimal Amount { get; set; }

    public Expense? Expense { get; set; }

    public DateTimeOffset CreatedAtUtc { get; set; } = DateTimeOffset.UtcNow;
}

/// <summary>Categories used when planning a trip's expected costs.</summary>
public enum PlannedCostCategory
{
    TravelToFrom,
    Accommodation,
    LocalTransport,
    Food,
    ActivitiesAndMuseums,
    BarsAndNightlife,
    Shopping,
    EmergencyBuffer,
    Other
}
